src/lib.rs

```diff
@@ -23,212 +21,95 @@
-  assert!(tt_iter.next().is_none(), USAGE);
-
-  let mut lit_string = format!("{}", lit);
-  let is_raw = lit_string.starts_with("r");
-  if is_raw {
-    lit_string.remove(0);
-    while lit_string.chars().nth(0) == Some('#') {
-      lit_string.pop();
-      lit_string.remove(0);
-    }
-  }
-
-  // right now we only support double quoted strings
-  assert!(lit_string.as_bytes().first() == Some(&b'"'), USAGE);
-  assert!(lit_string.as_bytes().last() == Some(&b'"'), USAGE);
-  let lit_str = &lit_string[1..lit_string.len() - 1];
-
-  str_to_utf16_units_tokenstream(lit_str, is_raw)
-}
-
-/// Turns a string literal into a `&[u16]` literal with a null on the end.
-///
-/// If you do **not** want to have a null terminator added to the string then
-/// you should use [`utf16!`](utf16!).
-#[proc_macro]
-pub fn utf16_null(stream: TokenStream) -> TokenStream {
-  const USAGE: &str = "Usage: utf16!(string_lit)";
-
-  let mut tt_iter = stream.into_iter();
-  let lit = match tt_iter.next().expect(USAGE) {
-    TokenTree::Literal(lit) => lit,
-    _ => panic!(USAGE),
-  };
-  // we expect only one string literal per invocation.
-  assert!(tt_iter.next().is_none(), USAGE);
-
-  let mut lit_string = format!("{}", lit);
-  let is_raw = lit_string.starts_with("r");
-  if is_raw {
-    lit_string.remove(0);
-    while lit_string.chars().nth(0) == Some('#') {
-      lit_string.pop();
-      lit_string.remove(0);
-    }
-  }
-
-  // right now we only support double quoted strings
-  assert!(lit_string.as_bytes().first() == Some(&b'"'), USAGE);
-  assert!(lit_string.as_bytes().last() == Some(&b'"'), USAGE);
-  // we need a null on the end, so we just reuse the end of this string.
-  lit_string.pop();
-  lit_string.push('\0');
-  let lit_str = &lit_string[1..];
-
-  str_to_utf16_units_tokenstream(lit_str, is_raw)
-}
-
-fn str_to_utf16_units_tokenstream(s: &str, is_raw: bool) -> TokenStream {
-  let mut encode_buf = [0_u16; 2];
-  let mut buf = String::with_capacity(s.as_bytes().len() * 8 + 10);
-  //
-  buf.push_str("&[");
-
-  if is_raw {
-    for c in s.chars() {
-      for unit in c.encode_utf16(&mut encode_buf) {
-        let _cant_fail = write!(buf, "{},", unit);
-      }
-    }
-  } else {
-    for char_escape in CharEscapeIterator::new(s.chars()) {
-      match char_escape {
-        CharEscape::Escaped(ch) | CharEscape::Literal(ch) => {
-          for unit in ch.encode_utf16(&mut encode_buf) {
-            let _cant_fail = write!(buf, "{},", unit);
-          }
-        }
-        other => panic!("Illegal character escape sequence: {:?}", other),
+macro_rules! imp {
+  ($(
+    $(#[$m:meta])*
+    $name:ident has $n:literal trailing zeroes
+  )*) => {
+    $(
+      $(#[$m])*
+      #[macro_export]
+      macro_rules! $name {
+        ($text:expr) => {{
+          // Try to avoid collisions with `text`'s scope
+          const __SWEIRFOH2387OPC: &str = $text;
+          const UTF8: &str = __SWEIRFOH2387OPC;
```

Comment:
I don't understand why you don't just assign `$text` directly to `UTF8`

Reply:
When the macro expands, it makes the `const` visible to the `$text` expansion. If we were to simply use `UTF8`, users wouldn't be able to write this (it complains about a circular dependency on itself):

```
const UTF8: &str = "foobar";
const UTF16: &[u16] = utf16!(UTF8);
```

How could I make the comment explaining this clearer?

Reply:
Okay, I understand what's going on more now.

I think the comment could be:
```rust
// Here we pick a name highly unlikely to exist in the scope that $text came from, which prevents a potential const eval cycle.
```